Update rent charge amounts

rf-mt opened the pull request:
Add end point which receives year and rc uplift
Add use case which creates new row for year and uplift, and creates new rows, per rent charge, for the new year
Add gateways for the fixed data and rent charge data

Add corresponding tests

rf-mt committed:
Add controller action, use case and gateways for
updating rent charges according to the received year
and rc uplift
Add corresponding tests

rf-mt committed:
Updated gateway tests

rf-mt committed:
Update request tests

rf-mt committed:
Drop cpi_fixed_data and rent_summaries tables
which appeared after rebasing master into this branch
and running migrations

rf-mt committed:
Update naming convention to rent charges
rather than rent details for consistency

rf-mt committed:
Add more tests to request spec

rrpff commented on lib/gateways/fixed_data_gateway.rb:
```diff
@@ -0,0 +1,9 @@
+  end
+
+  def all
+    FixedDatum.all
```
When using Rails' ActiveRecord from a gateway, I generally like to return a non-ActiveRecord object, as it prevents the rest of the system from being polluted by Rails. As this stands, the use case will receive an ActiveRecord object. This risks other parts of the system calling Rails model methods, breaking the gateway/use-case boundary.

I'd recommend creating a domain object called Domain::FixedDatum or similar, and mapping the model objects returned by FixedDatum.all into those domain objects. 

Shortened example of this from my WIP:

```
module Domain
  class Rent
    attr_accessor :uprn,
      :address,
      :address_2,
      ...

    def initialize(**config)
      config.each do |key, value|
        self.send("#{key}=", value)
      end
    end

    # example of how we can add dynamic attributes
    def lower_of_formula_rent_or_cap
      [formula_rent_this_year, rent_cap_this_year].min
    end
  end
end
```

rf-mt replied:
I haven't created a domain object for the fixed data because this PR only saves the fixed data, rather than returning it, therefore I don't actually need the `all` method in`FixedDataGateway`. I've created a domain object for `RentCharge` though

rrpff commented on lib/gateways/fixed_data_gateway.rb:
```diff
@@ -0,0 +1,9 @@
+class FixedDataGateway
+  def save(fixed_data)
+    FixedDatum.create(fixed_data)
```
This could fail silently at the moment, as `#create` won't riase an error if `fixed_data` isn't valid. You can use `#create!` to have it raise an error instead. It might be worth moving the validation from the model to the use case, as this allows validation which works the same regardless of how we're persisting the data. e.g. if we switched from ActiveRecord to posting to an API.

rrpff commented on lib/gateways/rent_charges_gateway.rb:
```diff
@@ -0,0 +1,11 @@
+  def all(year)
+    RentCharge.where(year: year)
+  end
+end
```
Same feedback here.

rrpff commented on lib/use_cases/update_rent_charges.rb:
```diff
@@ -0,0 +1,53 @@
+  def execute(fixed_data)
+    @fixed_data = fixed_data
+    @rc_uplift = @fixed_data[:rc_uplift].to_f
+    @year = @fixed_data[:year].to_i
```
I suggest changing these to normal variables instead of instance variables, as this means the use case is currently stateful. Meaning if we called execute more than once, state will persist within the use case instance between them. It won't actually cause any problems now, but potentially risky if we were to add conditional logic or anything more complicated in here.

rrpff commented on lib/use_cases/use_case_factory.rb:
```diff
@@ -0,0 +1,13 @@
+    UpdateRentCharges.new(fixed_data_gateway: fixed_data_gateway, rent_charges_gateway: rent_charges_gateway)
+  end
+
+  def fixed_data_gateway
```
Would recommend making the `*_gateway` methods private, as controllers shouldn't have access to them directly. Only the use cases need them.

rrpff commented on spec/requests/rent_charges_spec.rb:
```diff
@@ -0,0 +1,118 @@
+      expect(response).to have_http_status(201)
+    end
+  end
+end
```
Happy path tests look good in here. Suggest we add a couple for invalid inputs as well, so we can show basic error handling.

rrpff replied:
Nice changes 👍

rrpff commented on spec/unit/lib/gateways/fixed_data_gateway_spec.rb:
```diff
@@ -0,0 +1,12 @@
+    expect(fixed_data[0][:rc_uplift]).to eq(0.00015)
+    expect(fixed_data[0][:year]).to eq(2011)
+  end
+end
```
I recommend adding an additional example for triangulation, as at the moment the FixedDataGateway could be slimed - containing hardcoded 0.00015 and 2011 values.

rrpff commented on spec/unit/lib/gateways/rent_charges_gateway_spec.rb:
```diff
@@ -0,0 +1,74 @@
+    expect(rent_charges[1][:rent_cap_this_year]).to eq(120)
+    expect(rent_charges[1][:uprated_actual]).to eq(110)
+  end
+end
```
Same problem with triangulation here. I like to use https://github.com/stympy/faker sometimes to generate random test data, alongside a handwritten example.

The handwritten one proves the test output is coming back exactly as we want it. The randomised one provides confidence that we're not hardcoding anything.

rrpff commented on spec/support/request_spec_helper.rb:
```diff
@@ -0,0 +1,5 @@
+module RequestSpecHelper
+  def json
```
Looks good, but concerned the method name doesn't obviously describe what this does. It could be misinterpreted as serialising json from a hash, for example. `json_response` ?

rrpff reviewed (commented):
Looks good! Fair bit of feedback here, but mostly so we can standardise around some design decisions now. Let me know your thoughts 👍

rf-mt committed:
Add domain object for rent charges
Add validation for year and rc uplift
Use random test data in gateway and use case tests

rf-mt committed:
Add random test data to request spec

rrpff commented on lib/domain/rent_charge.rb:
```diff
@@ -0,0 +1,25 @@
+                  :rent_cap_this_year,
+                  :uprated_actual
+    
+    def initialize(**rent_charges)
```
`rent_charges` here is a hash of data for a single rent charge, while the variable name suggests an array of rent charge objects to me. Is there a better name we could use?

rf-mt replied:
good point, the only thing I can think of right now is `rent_charge`, but feel like I'm using this a lot. Is it convention to use `config` like you have in your example above?

rrpff commented on lib/gateways/fixed_data_gateway.rb:
```diff
@@ -0,0 +1,5 @@
+class FixedDataGateway
+  def save(fixed_data)
+    FixedDatum.create!(fixed_data)
```
Given we don't intend to use the value returned by this, I would explicitly return `nil` to avoid leaking ActiveRecord objects.

rrpff commented on lib/use_cases/update_rent_charges.rb:
```diff
@@ -0,0 +1,99 @@
+  
+    rent_charges_for_given_year = @rent_charges_gateway.all((fixed_data[:year].to_i) - 1)
+    @rent_charges_for_given_year = rent_charges_for_given_year.map do |rent_charge|
+      Domain::RentCharge.new(
```
I wouldn't construct the Domain::RentCharge object here, for a couple of reasons:

a) It's only used within this file currently, where it is immediately mapped into another hash by `rent_charges`, so I don't think it's providing a lot of value?
b) The main benefit of using Domain Objects comes when returning them from gateways. It allows all gateways which adhere to a responsibility/interface (e.g. SqlRentChargeGateway, ApiRentChargeGateway, InMemoryRentChargeGateway) to return a standardised object with behaviour that's useful to the use case that receives them. 

I'd suggest constructing the Domain::RentCharge object in the RentChargesGateway and returning it from the `#all` method, rather than using it in use case.

Correct me if I'm missing anything!

rf-mt replied:
yep sounds good!

rrpff commented on lib/gateways/rent_charges_gateway.rb:
```diff
@@ -0,0 +1,11 @@
+class RentChargesGateway
+  def save(rent_charges)
+    rent_charges.each do |rent_charge|
+      RentCharge.create!(rent_charge)
```
This has become slightly ambiguous to read now, as we have two `RentCharge` objects - the Domain one, and the ActiveRecord one. We could namespace all models so this was `Models::RentCharge`, for example?

rrpff commented on lib/use_cases/update_rent_charges.rb:
```diff
@@ -0,0 +1,99 @@
+      errors.push(:"missing_#{key}") if fixed_data[key].blank?
+    end
+    errors
+  end
```
I like this a lot, but think we're now duplicating validation between the use case and the model definition. Would recommend keeping it in the use case, and removing it from the model definition. That way, our validation isn't specific to Rails, and could be used to other gateways. Will also mean we're unlikely to have ActiveRecord errors raised, which neatens up our error handling!

rf-mt replied:
yes makes sense!

rrpff commented on spec/unit/lib/gateways/rent_charges_gateway_spec.rb:
```diff
@@ -0,0 +1,114 @@
+
+  let(:rent_charges_for_different_year) { [
+    { 
+      uprn: Faker::Base.regexify(/[0-9]{3}[a-z]{3}/),
```
I've never used `#regexify` - pretty cool!

rrpff commented on spec/unit/lib/gateways/rent_charges_gateway_spec.rb:
```diff
@@ -0,0 +1,114 @@
+      removed: :false,
+      formula_rent_this_year: Faker::Number.between(100, 200),
+      rent_cap_this_year: Faker::Number.between(100, 200),
+      uprated_actual: Faker::Number.between(100, 200)
```
No randomised data for rr_count, property_type, year or removed?

Should removed be a symbol or a boolean?

rf-mt replied:
I was having issues with property_type because it's an enum. I was doing `Faker::Number.between(0,4)` for the test data but the tests were failing because the value pulled from the db is the actual enumerated value. I couldn't figure out how to test it using random data?

rf-mt replied:
and yeah `removed` should be a boolean (updated). I was struggling to randomise `year` and retain control over the tests earlier (less so in the rent charge gateway test, but more in the update rent charge unit tests and request spec), but think I've sorted it now (also updated).

rrpff replied:
Nice. You could consider storing the enum in a constant like Models::RentCharge::PROPERTY_TYPES, then using #sample to grab a random type.

rf-mt replied:
ah yes, thanks!

rrpff commented on spec/unit/lib/use_cases/update_rent_charges_spec.rb:
```diff
@@ -0,0 +1,113 @@
+    end
+  end
+
+  it 'raises error when missing year' do
```
Might use terminology like `returns an error` or `includes an error` rather than `raises an error`. Super nitpicky, but `raise` has very specific connotations.

rf-mt replied:
makes sense!

rrpff reviewed (commented):
Looks great, feeling confident about this. Have added a few more recommendations.

rf-mt committed:
Add namespace for models
Construct Domain::RentCharge object in gateway rather than usecase
Add randomised data for rr_count and removed

jpmaHub commented on app/controllers/application_controller.rb:
```diff
@@ -1,4 +1,9 @@
 class ApplicationController < ActionController::API
+  include Response
+
+  def rent_charges_use_case_factory
```
Looks like this method could be reusable in `view historical percent changes` controller too. So, I think it is better to use a generic name that can be used by other controllers.

rf-mt replied:
sure, I've just renamed it to `use_case_factory` for now

rf-mt committed:
Add randomised test data for year in request spec,
rent charge gateway tests and update rent charge unit tests

rrpff commented on lib/gateways/rent_charges_gateway.rb:
```diff
@@ -0,0 +1,30 @@
+        uprated_actual: rent_charge[:uprated_actual]
+      )
+    end
+    rent_charges_by_year
```
Could just return the `Models::RentCharge.where(...).map { ... }` value without assigning it to a variable. Very slightly neater.

rf-mt replied:
yep! definitely neater

rrpff commented on lib/use_cases/update_rent_charges.rb:
```diff
@@ -0,0 +1,78 @@
+        bedroom_weight: rent_charge.bedroom_weight,
+        mra_archetype: rent_charge.mra_archetype,
+        jan_1999_asset_values: rent_charge.jan_1999_asset_values,
+        year: rent_charge.year + 1,
```
Why do we have to add 1 here?

rf-mt replied:
adding 1 because here we are building the rent charges for the next year, so we pull out the rent charges for the previous year and then perform the necessary calculations to update (including incrementing the year)

rrpff commented on lib/use_cases/update_rent_charges.rb:
```diff
@@ -0,0 +1,78 @@
+  private
+
+  def rent_charges(fixed_data)
+    rent_charges = @rent_charges_gateway.all((fixed_data[:year].to_i) - 1)
```
Why do we have to take away 1 here?

rf-mt replied:
so similar to my previous comment, the endpoint receives the new year for which the user wants to update the rent charges, so we need to pull the previous years rent charges to perform the calculations on to create the new year's charges

rrpff commented on spec/requests/rent_charges_spec.rb:
```diff
@@ -0,0 +1,140 @@
+        formula_rent_this_year: Faker::Number.between(100, 200),
+        rent_cap_this_year: Faker::Number.between(100, 200),
+        uprated_actual: Faker::Number.between(100, 200)
+      }
```
Given the three of these are very similar, you could define a method to generate these, rather than duplicating it. A pattern I often use it to define a method which takes a hash as an argument, and merges it into a static hash of default data. That way in your tests you can set up data like:

```
[
  generate_rent_charge(year: year),
  generate_rent_charge(year: year - 25),
  # etc
]
```

Especially useful when you want to test the effects of a specific attribute, without reducing the clarity of the test by including all the other attributes.

rrpff commented on spec/requests/rent_charges_spec.rb:
```diff
@@ -0,0 +1,140 @@
+        expect(rent_charges_gateway.all(rent_charges.first[:year] + 1).second.year).to eq(rent_charges.first[:year] + 1)
+        expect(rent_charges_gateway.all(rent_charges.first[:year] + 1).second.formula_rent_this_year).to eq((rent_charges.third[:formula_rent_this_year] * 0.99).round(2))
+        expect(rent_charges_gateway.all(rent_charges.first[:year] + 1).second.rent_cap_this_year).to eq((rent_charges.third[:rent_cap_this_year] * 0.99).round(2))
+        expect(rent_charges_gateway.all(rent_charges.first[:year] + 1).second.uprated_actual).to eq((rent_charges.third[:uprated_actual] * 0.99).round(2))
```
Similar feedback here - you could perhaps use a loop to test both?

rrpff reviewed (approved):
Looks great, the last bits of feedback here are more on style/refactoring. This looks ready to go.

rf-mt committed:
Refactor request spec and usecase spec
